manubot/cite/util.py

```diff
@@ -244,18 +249,96 @@ def citation_to_citeproc(citation, prune=True):
+    for prefix in prefixes:
+        if citation.startswith(prefix):
+            return citation
+        if citation.lower().startswith(prefix):
+            return prefix + citation[len(prefix):]
+    return f'raw:{citation}'
+
+
+def csl_item_set_standard_citation(csl_item):
+    """
+    Extract the standard_citation for a csl_item and modify the csl_item in-place to set its standard_citation.
+    The standard_citation is extracted from a "standard_citation" field, the "note" field, or the "id" field.
+    If extracting the citation from the "id" field, use the infer_citation_prefix function to set the prefix.
+    For example, if the extracted standard_citation does not begin with a supported prefix (e.g. "doi:", "pmid:"
+    or "raw:"), the citation is assumed to be raw and given a "raw:" prefix. The extracted citation
+    (referred to as "original_standard_citation") is checked for validity and standardized, after which it is
+    the final "standard_citation".
+
+    Regarding csl_item modification, the csl_item "id" field is set to the standard_citation and the note field
+    is created or updated with key-value pairs for standard_citation, original_standard_citation, and original_id.
+    """
+    if not isinstance(csl_item, dict):
+        raise ValueError("csl_item must be a CSL Data Item represented as a Python dictionary")
+
+    from manubot.cite.citeproc import (
+        append_to_csl_item_note,
+        parse_csl_item_note,
+    )
+    note_dict = parse_csl_item_note(csl_item.get('note', ''))
+
+    original_id = None
+    original_citation = None
+    if 'id' in csl_item:
+        original_id = csl_item['id']
+        original_citation = infer_citation_prefix(original_id)
+    if 'standard_citation' in note_dict:
+        original_citation = note_dict['standard_citation']
+    if 'standard_citation' in csl_item:
+        original_citation = csl_item.pop('standard_citation')
+    if original_citation is None:
+        raise ValueError(
+            'csl_item_set_standard_citation could not detect a field with a citation / standard_citation. '
+            'Consider setting the CSL Item "id" field.'
+        )
+    assert is_valid_citation_string('@' + original_citation, allow_raw=True)
+    standard_citation = standardize_citation(original_citation, warn_if_changed=False)
+    add_to_note = {}
+    if original_id and original_id != standard_citation:
+        if original_id != note_dict.get('original_id'):
+            add_to_note['original_id'] = original_id
+    if original_citation and original_citation != standard_citation:
+        if original_citation != note_dict.get('original_standard_citation'):
+            add_to_note['original_standard_citation'] = original_citation
+    if standard_citation != note_dict.get('standard_citation'):
+        add_to_note['standard_citation'] = standard_citation
+    append_to_csl_item_note(csl_item, dictionary=add_to_note)
```

Comment:
If the note already has the key and the value hasn't changed, it will not be appended again. However, if the value has changed, the key-value pair will be appended. At least, the note parser we use here should take use the value from the last key if reading in again.

A more complex option would be a way to update an exiting key or write a new key, but I didn't think this corner case was worth the complexity. We will see if this becomes an issue in practice.

Reply:
I agree. The behavior you describe is reasonable, and the added complexity would not be worthwhile.